### ISSUE TITLE
Change in behavior when copying a LasHeader object

### ISSUE THREAD

**Issue by AlexBass05** [PDAL/PDAL#4008]
I'm encountering a change in behavior when copying a `LasHeader` object.

Consider the following unit test
```
TEST(LasReaderTest, header_modified)
{
    auto getHeader = [](const std::string& path){
        PointTable table;
        Options ops;
        ops.add("filename", path);

        LasReader reader;
        reader.setOptions(ops);

        reader.prepare(table);
        return reader.header();
    };

    LasHeader h = getHeader(Support::datapath("las/simple.las"));

    EXPECT_EQ(h.softwareId(), "TerraScan");

   // Remaining checks omitted
}
```

This test is a modified version of the existing ["LasReaderTest, header"](https://github.com/PDAL/PDAL/blob/master/test/unit/io/LasReaderTest.cpp#L120) with the difference that the `LasReader` goes out of scope after creating a copy of the `LasHeader` and the added `EXPECT_EQ` check.

**Expected behavior**
This test passes. (That is at least my expectation)

**Actual behavior**
The actual behavior depends on the PDAL version:
* `PDAL 2.3.0`: The test passes.
* `PDAL 2.5.2/master`: The test fails with
```
Expected equality of these values:
  h.softwareId()
    Which is: "\f\xEDJ\xECtU\0\0M"
  "TerraScan"
```

To me this looks like we are accessing an invalid memory location because the copy-constructor of `LasHeader` is "only" making a shallow copy with some reference pointing to the `LasReader`-object.

I believe the change in behavior was introduced in https://github.com/PDAL/PDAL/commit/348ffbf4db8f46f3179e7ac46eaadb404cf1f06f (shortly before `PDAL 2.4.0`).
Note that one commit before (https://github.com/PDAL/PDAL/commit/2d846ac8d38020703379e40b548c6eab05586f8b) the above test does not compile because `LasHeader` is missing a copy-constructor.

As a client of this API it is not clear that the copy-constructor of `LasHeader` is not making a deep copy.

**System/installation information:**
Tested with building PDAL from source on the above mention git tags/commits on Ubuntu 22.04.

**Comment by hobu** [comment]
This was a consequence of trying to preserve the old API https://lists.osgeo.org/pipermail/pdal/2022-January/002525.html

**Comment by hobu** [comment]
https://github.com/PDAL/PDAL/pull/3682

**Comment by abellgithub** [comment]
To perhaps clarify... the data in LasHeader is stored in the reader, so you must keep the LasReader alive for the LasHeader to stay valid. When we added COPC support, we made EPT/COPC/LAS header information private and accessible as metadata, which is how data for all other readers is accessible. The LasHeader interface has been maintained for backward compatibility but requires the LasReader from which the `header()` is derived to be retained.

**Comment by hobu** [comment]
>  the data in LasHeader is stored in the reader, so you must keep the LasReader alive for the LasHeader to stay valid.

Should we back-out these copy constructors? The current behavior is error-prone at the very least.

**Comment by abellgithub** [comment]
There's nothing wrong with copying. All copies reference the data in the
reader. The copy support was also necessary to maintain backward
compatibility for some users. It is error prone. Perhaps document better?

**Comment by AlexBass05** [comment]
Apologies for the late reply.

> It is error prone.

We all seem to agree on that. 

> Perhaps document better?

Yes, some documentation would help.

The suggested fix in #4011 looks like a good alternative to me. I cannot judge the implications with respect to backward compatibility though.
